src/app/data/types/IInventoryArmor.ts

```diff
@@ -30,5 +37,77 @@ export interface IInventoryArmor extends IManifestArmor {
   intellect: number;
   strength: number;
   energyLevel: number;
+
+  // Note: this will be empty for vendor items
   statPlugHashes: (number | undefined)[];
+
+  source: InventoryArmorSource;
+}
+
+export function createArmorItem(
+  manifestItem: IManifestArmor,
+  itemInstanceId: string,
+  source: InventoryArmorSource
+): IInventoryArmor {
+  if (!manifestItem) throw new Error("Missing manifest item");
+
+  const item: IInventoryArmor = Object.assign(
+    {
+      id: -1,
+      itemInstanceId,
+      mayBeBugged: false,
+      statPlugHashes: [],
+      masterworked: false,
+      energyLevel: 0,
+      mobility: 0,
+      resilience: 0,
+      recovery: 0,
+      discipline: 0,
+      intellect: 0,
+      strength: 0,
+      source,
+    },
+    manifestItem
+  );
+  (item as any).id = undefined;
+
+  // HALLOWEEN MASKS
+  if (
+    manifestItem.hash == 2545426109 ||
+    manifestItem.hash == 199733460 ||
+    manifestItem.hash == 3224066584
+  ) {
+    item.slot = ArmorSlot.ArmorSlotHelmet;
+  }
+
+  return item;
+}
+
+// This will mutate the incoming armor item to add the relevant stats
+// plugHashes is received as a simple array of hashes,
+// as the data is in a different shape for instances vs manifest items
+export function applyInvestmentStats(
+  r: IInventoryArmor,
+  investmentStats: { [id: number]: number }
+) {
+  r.mobility = investmentStats[2996146975];
+  r.resilience = investmentStats[392767087];
+  r.recovery = investmentStats[1943323491];
+  r.discipline = investmentStats[1735777505];
+  r.intellect = investmentStats[144602215];
+  r.strength = investmentStats[4244567218];
+}
+
+// Returns true if the items are effectively equal in stats
+export function isEqualItem(a: IInventoryArmor, b: IInventoryArmor): boolean {
+  return (
```

Comment:
In theory, items also differ in "has perk/slot or not" (e.g. Artifice). The comment states `true if the items are effectively equal in stats`, so ignore this if you don't think it is relevant.

Reply:
I think that the available sockets are only defined by the manifest item, which we are comparing here, so I believe this should be safe?